### PR TITLE
Ccy pairs showing on screen

### PR DESCRIPTION
**Pull request by JakeGinnivan**
And a bunch more work on services

### CODE REVIEW

**Comment by JakeGinnivan** [comment]
@raybooysen this builds on your other PR and gets a few things working.

**Comment by JakeGinnivan** [comment]
@raybooysen pushed some stuff. Unsure how much ill do on the next flight, but probably a little bit

File: src/components/tiles/spotTile.jsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+      }
+    })
+  }
+
+  render() {
+    if (this.state.isStale) {
+      return <StaleTile />
+    }
+
+    // todo should ccypair be context?
```
Comment (raybooysen) [inline src/components/tiles/spotTile.jsx]: I think this makes sense.  It should propogate the entire way down.

File: src/components/tiles/spotTiles.jsx [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      isLoading: true,
+      currencyPairs: []
+    }
+  }
+
+  componentDidMount() {
+    referenceDataService.getCurrencyPairUpdatesStream()
+      .subscribe(u => {
+        this.props.loaded()
+        console.log('Spot tiles loaded');
```
Comment (raybooysen) [inline src/components/tiles/spotTiles.jsx]: This yields n times, we need to handle this. Better logging etc.

File: src/components/tiles/spotTiles.jsx [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        })
+      })
+  }
+
+  render() {
+    return (
+      <div style={{overflow: 'auto'}}>
+        <ul style={{listStyle: 'none', padding: 0}}>
+          {this.state.currencyPairs.map(c => (
+            <li className='spot-tile' key={c.symbol}><SpotTile ccyPair={c} /></li>)
```
Comment (raybooysen) [inline src/components/tiles/spotTiles.jsx]: Did we support duplicate tiles (same pair) in the old reactive trader?  I can't remember.

File: src/components/tiles/states/pricingTile.jsx [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    this.execute = this.execute.bind(this);
+  }
+
+  formatPrice(rate) {
+    return PriceFormatter.getFormattedPrice(rate,
+      this.props.ccyPair.ratePrecision, this.props.ccyPair.pipsPosition)
+  }
+
+  execute(direction, rate) {
+    return this.props.onExecute(direction, rate, this.props.price.valueDate)
```
Comment (raybooysen) [inline src/components/tiles/states/pricingTile.jsx]: Can this be handled differently as a pattern?  For example, if value date is already a prop, then why does the onExecute (which is also a prop) need to know about it?

File: src/components/tiles/states/pricingTile.jsx [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  }
+
+  render() {
+    let prices;
+    if (this.props.price) {
+      prices = (
+        <div>
+          <OneWayPrice direction='sell'
+                       formattedPrice={this.formatPrice(this.props.price.bid)}
```
Comment (raybooysen) [inline src/components/tiles/states/pricingTile.jsx]: Not for this PR, but we should move the formatPrice() up into the stream. Once you have n tiles sharing the same stream, this becomes a hotspot.

File: src/helpers/rxExtensions.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+          scheduleStale()
+
+          return new Rx.CompositeDisposable(sourceSubscription, timerSubscription)
+      })
+    }
+  }
+}
+
+// TODO figure out how to actually export properly
```
Comment (raybooysen) [inline src/helpers/rxExtensions.js]: What do you mean?

File: src/main.jsx [hunk 7]
```diff
@@ -1,20 +1,53 @@
...
+import Analytics from 'components/analytics/analytics.jsx'
+import Tiles from 'components/tiles/spotTiles.jsx'
+import Blotter from 'components/blotter/blotter.jsx'
 
 export default class Main extends React.Component {
   constructor(props) {
-    super(props);
-
+    super(props)
+    this.state = {
```
Comment (raybooysen) [inline src/main.jsx]: these flags should be a function of the components, not the owning component.

File: src/model/direction.js [hunk 8]
```diff
@@ -1,6 +0,0 @@
-const Direction = Object.freeze(
-  {
-    Buy: Symbol('Buy'),
```
Comment (raybooysen) [inline src/model/direction.js]: how come this has been deleted?

File: src/model/pricing/executablePrice.js [hunk 9]
```diff
@@ -1,11 +1,14 @@
...
 
-    execute(notional, dealtCurrency) {
-        return this._executionRepository.executeRequest(this, notional, dealtCurrency);
-    }
-}
\ No newline at end of file
+export default AmpersandState.extend({
+  extraProperties: extraProperties(),
+  props: {
+    direction: { type: 'string', required: true, values: ['bid', 'ask'] },
```
Comment (raybooysen) [inline src/model/pricing/executablePrice.js]: we've now got the direction duplicated.

File: src/model/pricing/price.js [hunk 10]
```diff
@@ -1,16 +1,20 @@
...
 
-    bid.parent = this;
-    ask.parent = this;
-
-    this.spread = (ask.rate - bid.rate) * Math.pow(10, currencyPair.pipsPosition);
-    this.mid = (this.bid.rate + this.ask.rate) / 2;
+  derived: {
+    spread: {
+      deps: ['bid', 'ask', 'currencyPair'],
+      fn: () => (this.ask.rate - this.bid.rate) * Math.pow(10, this.currencyPair.pipsPosition)
```
Comment (raybooysen) [inline src/model/pricing/price.js]: nice. :)

File: src/services/extraProperties.js [hunk 11]
```diff
@@ -0,0 +1,10 @@
+import isProduction from './isProduction.js'
+
+export default function() {
+  let extraProperties = 'reject'
+  if (isProduction()) {
+    extraProperties = 'ignore'
```
Comment (raybooysen) [inline src/services/extraProperties.js]: What does this do?

File: src/services/serviceClients/signalR/transport/connection.js [hunk 12]
```diff
@@ -0,0 +1,141 @@
+import Rx from 'rx'
+import ConnectionType from './connectionType.js'
+import ConnectionInfo from './connectionInfo.js'
+import ConnectionStatus from './connectionStatus.js'
+import $ from 'jquery'
+import 'ms-signalr-client'
+
+export default class Connection {
+    constructor(address, username) {
+      this._status = new Rx.BehaviorSubject(new ConnectionInfo(ConnectionStatus.Uninitialized, address, ConnectionType.None))
```
Comment (raybooysen) [inline src/services/serviceClients/signalR/transport/connection.js]: I kinda like the _ for convention based private methods.  Happy to follow this?

**Comment by raybooysen** [comment]
Direction looks good. :)

**Comment by raybooysen** [comment]
Also, as a bigger picture thing, we need to think about versioning of prices to ensure we can check whether we need to reender or not.

File: src/components/tiles/states/pricingTile.jsx (hunk 4 above)
Comment (JakeGinnivan) [inline src/components/tiles/states/pricingTile.jsx]: Good point. That is what I was going for. Curry the execute methods up the chain? Good pattern or not?

File: src/components/tiles/states/pricingTile.jsx (hunk 5 above)
Comment (JakeGinnivan) [inline src/components/tiles/states/pricingTile.jsx]: Yeah, the levels of these components are not quite right yet. Will adjust once entire stack is in

File: src/helpers/rxExtensions.js (hunk 6 above)
Comment (JakeGinnivan) [inline src/helpers/rxExtensions.js]: I tried `export function foo() { }` but foo is not exported. Unsure why i have to create an object and export that. Need to read up. Or is this right?

File: src/main.jsx (hunk 7 above)
Comment (JakeGinnivan) [inline src/main.jsx]: I want to show the loading for the app until at least the main components are initialised. RT shows the splash screen then it loads in one go

File: src/model/direction.js (hunk 8 above)
Comment (JakeGinnivan) [inline src/model/direction.js]: Using the ampersand model which enforces either 'buy' or 'sell' as a string, not using enums just well known strings

File: src/model/pricing/executablePrice.js (hunk 9 above)
Comment (JakeGinnivan) [inline src/model/pricing/executablePrice.js]: Yeah, noticed that. Will try sort out bit later

File: src/services/extraProperties.js (hunk 11 above)
Comment (JakeGinnivan) [inline src/services/extraProperties.js]: If there are properties on the dtos from the server the model blows up. We don;t want that happening in prod but useful for development

File: src/services/serviceClients/signalR/transport/connection.js (hunk 12 above)
Comment (JakeGinnivan) [inline src/services/serviceClients/signalR/transport/connection.js]: Yeah, _ is good

**Comment by JakeGinnivan** [comment]
Still needs styling work and still not executing, but tiles are taking shape

File: src/components/tiles/spotTiles.jsx (hunk 3 above)
Comment (LeeCampbell) [inline src/components/tiles/spotTiles.jsx]: Nope. Duplicate tiles not supported in the demos i give

**Pull request merged**